### PR TITLE
Link level `NeighborLoader`

### PR DESCRIPTION
**Pull request by Padarn**
Adds a link neighbourhood loader that is intended to provide utility for sampling for link prediction tasks.

The current API looks like: 
```
loader = LinkNeighborLoader(normal_data, num_neighbors=[2, 1], batch_size=1, shuffle=False)
data, target_edge_index, target_labels = next(iter(loader))
```
The loader iterates giving the sampled g raph and the target edge index and labels for the edges in the graph. The labels are assumed to be "True" for all edges in the graph.

TODO 
- [x] Add hetro graph compatability
- [x] Cleanup hetro/non-hetro type checking
- [x] Examples
- [x] Tests

Addresses https://github.com/pyg-team/pytorch_geometric/issues/4026

### CODE REVIEW

File: torch_geometric/loader/link_neighbour_loader.py [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+
+        batch_size = len(index)
+
+        # take start and end node from each edge then deduplicate
+        row, col = self.data.edge_index
+        node_index = torch.cat([row[index], col[index]], dim=0)
+        node_index = torch.unique(node_index)
+
+        # get sampled graph
+        node, row, col, edge, _ = super().__call__(index)
```
Comment (rusty1s) [inline torch_geometric/loader/link_neighbour_loader.py:35]:
```suggestion
        node, row, col, edge, _ = super().__call__(node_index)
```

File: torch_geometric/loader/link_neighbour_loader.py [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+class LinkNeighborSampler(NeighborSampler):
+    def __init__(
+        self,
+        data: Data,
+        num_neighbors: NumNeighbors,
+        replace: bool = False,
+        directed: bool = True,
+    ):
+        super().__init__(data, num_neighbors, replace, directed)
+        self.data = data
```
Comment (rusty1s) [inline torch_geometric/loader/link_neighbour_loader.py:23]: It is pretty important to not save `data` here (as otherwise the data needs to be shared or replicated across `num_workers` during data loading). We could construct `node_index` directly from `self.colptr` and `self.row`.

File: torch_geometric/loader/link_neighbour_loader.py [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+        row, col = self.data.edge_index
+        node_index = torch.cat([row[index], col[index]], dim=0)
+        node_index = torch.unique(node_index)
+
+        # get sampled graph
+        node, row, col, edge, _ = super().__call__(index)
+        return node, row, col, edge, batch_size, index
+
+
+class LinkNeighborLoader(BaseDataLoader):
```
Comment (rusty1s) [inline torch_geometric/loader/link_neighbour_loader.py:39]:
```suggestion
class LinkNeighborLoader(torch.utils.data.DataLoader):
```
`BaseDataLoader` does no longer exist :(

File: torch_geometric/loader/link_neighbour_loader.py [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+    def __init__(
+        self,
+        data: Data,
+        num_neighbors: NumNeighbors,
+        input_edge_idx: InputEdges = None,
```
Comment (rusty1s) [inline torch_geometric/loader/link_neighbour_loader.py:44]:
```suggestion
        input_edges: InputEdges = None,
```
Aligns better with `NeighborLoader`. WDYT?

File: torch_geometric/loader/link_neighbour_loader.py [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+
+    def transform_fn(self, out: Any) -> Tuple[Data, torch.Tensor]:
+        node, row, col, edge, batch_size, index = out
+        data = filter_data(self.data, node, row, col, edge,
+                           self.neighbor_sampler.perm)
+
+        data.batch_size = batch_size
+        data = data if self.transform is None else self.transform(data)
+
+        labels = self.get_input_edge_labels()[index]
```
Comment (rusty1s) [inline torch_geometric/loader/link_neighbour_loader.py:80]: Shall we write this directly into `data`? (`edge_label` and `edge_label_index`)?

Comment (rusty1s) [inline torch_geometric/loader/link_neighbour_loader.py:80]: Any chance we only compute `get_input_edge_labels` and `get_input_edge_idx` once in `__init__`?

**Review by rusty1s — COMMENTED** [review]
This is a pretty good start. Really cool :)

Left some comments!

File: torch_geometric/utils/mask.py [hunk 6]
```diff
@@ -20,7 +20,7 @@ def index_to_mask(index: Tensor, size: Optional[int] = None) -> Tensor:
 
 
 def mask_to_index(mask: Tensor) -> Tensor:
-    r"""Converts mask to indicies representation 
+    r"""Converts mask to indicies representation
```
[the remark this answers is not in the document]
Comment (Padarn) [reply torch_geometric/utils/mask.py:23]: I need to set up spell checking in my IDE 😅

File: torch_geometric/loader/link_neighbour_loader.py (hunk 3 above)
Comment (Padarn) [reply torch_geometric/loader/link_neighbour_loader.py:39]: Oh I must need to rebase on master. Thanks for the heads up.

File: torch_geometric/loader/link_neighbour_loader.py (hunk 5 above)
Comment (Padarn) [reply torch_geometric/loader/link_neighbour_loader.py:80]: Yeah I was wondering about that. I thought it might be a little bit confusing to have a property that didn't exist on all edges, but its definitely much cleaner.

**Comment by Padarn** [comment]
Thanks, a lot of really good feedback. I'll address it all as I continue to work on the PR.

File: torch_geometric/loader/link_neighbour_loader.py (hunk 2 above)
Comment (Padarn) [reply torch_geometric/loader/link_neighbour_loader.py:23]: Good point!

File: torch_geometric/loader/link_neighbour_loader.py (hunk 3 above)
Comment (Padarn) [reply torch_geometric/loader/link_neighbour_loader.py:39]: Oh I see the change in https://github.com/pyg-team/pytorch_geometric/pull/4377 - I don't know much about lightning but its on my list to better understand what assumptions it is making about loaders :-)

File: torch_geometric/loader/neighbor_loader.py [hunk 7]
```diff
@@ -13,12 +13,14 @@
...
-from torch_geometric.typing import EdgeType, InputNodes
-
-NumNeighbors = Union[List[int], Dict[EdgeType, List[int]]]
+from torch_geometric.typing import InputNodes, NodeType, NumNeighbors
+from torch_geometric.utils import mask_to_index
 
 
 class NeighborSampler:
+
+    device = "cpu"
```
Comment (Padarn) [inline torch_geometric/loader/neighbor_loader.py:22]: I'm really not a fan of doing this, but it seemed a bit pointless to make it configurable, and I think its better than hard coding it in more places

File: torch_geometric/nn/models/deep_graph_infomax.py [hunk 8]
```diff
@@ -51,10 +51,12 @@ def discriminate(self, z, summary, sigmoid=True):
 
         Args:
             z (Tensor): The latent space.
+            summary (Tensor): The summary vector.
```
Comment (Padarn) [inline torch_geometric/nn/models/deep_graph_infomax.py:54]: Not totally sure why this one is here, it also seems to be in master.

File: torch_geometric/loader/link_neighbour_loader.py (hunk 4 above)
Comment (Padarn) [reply torch_geometric/loader/link_neighbour_loader.py:44]: The reason I changed this was that I need to eventually recover the index of the edges to look up any edge attributes that exist. So if taking as input the actual edges I need to do a search for each edge in the `data.edge_index`.

But... it does seem somehow more usable if you provide edges, so I'm a fan of your suggestion with the caveat that we'll need to do a mapping from edges -> edges_index.

Updating the PR with this, but let me know if you had something else in mind.

Comment (Padarn) [reply torch_geometric/loader/link_neighbour_loader.py:44]: Actually, I can't see an efficient way to do this... leaving both in the MR.. but I'm hesitant to keep the format without 
it explicitly being an index.

Comment (Padarn) [reply torch_geometric/loader/link_neighbour_loader.py]: Going to resolve this and move it as this file is gone (renamed)

File: torch_geometric/loader/link_neighbor_loader.py [hunk 9]
```diff
@@ -0,0 +1,146 @@
...
+            if isinstance(input_edges, EdgeType):
+                raise RuntimeError("`input_edges` cannot be string for non"
+                                   "hetrogenous graphs")
+
+            input_size = input_edges.size()
+
+            if len(input_size) == 1 and input_edges.dtype == torch.bool:
+                return mask_to_index(self.input_edge_idx)
+
+            if len(input_edges.size()) == 2 and input_edges.size()[0] == 2:
```
Comment (Padarn) [inline torch_geometric/loader/link_neighbor_loader.py:128]: Restart the conversation on the input specification here: The reason I initially changed the input name to `input_edge_idx` was so that it was clear we wouldn't do this (an inverse mapping from edges to their index). We need the index eventually because we need to recover the edge attributes in the transform function.

I'd be a fan of removing this personally and assuming its an index that is provided to this class - Although it would be nice to allow people to provide edges that don't exist and sample the graph for them, its also confusing what to do for the edge attributes in that case.

WDYT @rusty1s? Open to other approaches.

File: torch_geometric/loader/link_neighbour_loader.py (hunk 3 above)
Comment (rusty1s) [reply torch_geometric/loader/link_neighbour_loader.py]: There exists currently a bug in PL such that custom loaders need to directly inherit from `torch.utils.data.DataLoader`.

File: torch_geometric/loader/link_neighbour_loader.py (hunk 5 above)
Comment (rusty1s) [reply torch_geometric/loader/link_neighbour_loader.py]: I think this replicates our `RandomLinkSplit`quite well in which we have `data.edge_index` for message passing` and `data.edge_label_index` for supervision (with potentially different shape).

File: torch_geometric/loader/link_neighbor_loader.py (hunk 9 above)
Comment (rusty1s) [reply torch_geometric/loader/link_neighbor_loader.py:128]: I am not really sure I fully understand. Why do we need to recover edge attributes of supervision edges (that potentially do not even exist)?

In case you require an index to this loader, this means that you can only sample links that actually exist (i.e. which are "indexable" by `edge_index[idx]`), which does not seem to be the preferred solution. Ultimately, we want to infer missing links here as well.

File: torch_geometric/loader/neighbor_loader.py (hunk 7 above)
Comment (rusty1s) [reply torch_geometric/loader/neighbor_loader.py:22]: What's the reason for this change?

File: torch_geometric/nn/models/deep_graph_infomax.py (hunk 8 above)
Comment (rusty1s) [reply torch_geometric/nn/models/deep_graph_infomax.py:54]: Looks like something went wrong during rebase.

File: torch_geometric/loader/link_neighbor_loader.py (hunk 9 above)
Comment (Padarn) [reply torch_geometric/loader/link_neighbor_loader.py:102]: Got your point. I was being stupid.

I've updated the PR (it's still messy, but hopefully we're more aligned on the design). The only part I'm uncertain about is on line 85. Will open a conversation there.

File: torch_geometric/loader/neighbor_loader.py (hunk 7 above)
Comment (Padarn) [reply torch_geometric/loader/neighbor_loader.py:22]: I did this initially because I was storing something on CPU in the LinkNeighbourSampler too, so it seemed better to make sure they all shared the same device.. I'll revert

File: torch_geometric/loader/link_neighbour_loader.py (hunk 5 above)
Comment (Padarn) [reply torch_geometric/loader/link_neighbour_loader.py]: I've updated this to store `sampled_edges` and `sampled_edge_labels` (and `sampled_index`, maybe we can drop this one), so that it's easy to work with labels of edges that don't exist in the original graph.

File: torch_geometric/loader/link_neighbor_loader.py [hunk 10]
```diff
@@ -0,0 +1,108 @@
...
+
+        data.batch_size = batch_size
+        data = data if self.transform is None else self.transform(data)
+
+        data.sampled_index = index
+
+        edges = self.input_edges[:, index]
+        data.sampled_edges = edges
+
+        if self.input_edge_labels is not None:
```
Comment (Padarn) [inline torch_geometric/loader/link_neighbor_loader.py:85]: I wanted originally to return automatically the label as 'True' for existing edges and 'false' for non-existing ones, but doing this is a bit expensive (need to figure out which on the input edges are in the graph). Not sure its worth it.

Thoughts?

File: torch_geometric/nn/models/deep_graph_infomax.py (hunk 8 above)
Comment (Padarn) [reply torch_geometric/nn/models/deep_graph_infomax.py:54]: Yeah, I'll rebase and squash my PRs once the rest of the conversations are resolved

File: test/loader/test_neighbor_loader.py [hunk 11]
```diff
@@ -54,7 +54,7 @@ def test_homogeneous_neighbor_loader(directed):
         assert is_subset(batch.edge_index, data.edge_index, batch.x, batch.x)
 
         # Test for isolated nodes (there shouldn't exist any):
-        assert data.edge_index.view(-1).unique().numel() == data.num_nodes
+        assert batch.edge_index.view(-1).unique().numel() == batch.num_nodes
```
Comment (Padarn) [inline test/loader/test_neighbor_loader.py:57]: I assumed this was a mistake, can revert if not (otherwise same check is done for each batch)

File: torch_geometric/loader/link_neighbor_loader.py [hunk 12]
```diff
@@ -0,0 +1,307 @@
...
+                dim=0)
+            query_nodes, reverse_query_nodes = torch.unique(
+                query_nodes, return_inverse=True)
+
+            # get sampled graph
+            node, row, col, edge, _ = super().__call__(query_nodes)
+
+            # assume order of nodes in the new graph is same as input
+            assert all(query_nodes == node[:len(query_nodes)])
+            out_graph_nodes = torch.argsort(query_nodes)
```
Comment (Padarn) [inline torch_geometric/loader/link_neighbor_loader.py:70]: Any idea if there is a better way to do this? I do see https://discuss.pytorch.org/t/how-to-quickly-inverse-a-permutation-by-using-pytorch/116205 suggests a minor optimization, and there are a few on StackOverflow (I can benchmark some improvements in another MR), but I'm wondering if I'm missing a way to avoid this entirely.

File: test/loader/test_neighbor_loader.py (hunk 11 above)
Comment (rusty1s) [reply test/loader/test_neighbor_loader.py:57]: Thank you :)

File: torch_geometric/loader/link_neighbor_loader.py (hunk 12 above)
Comment (rusty1s) [reply torch_geometric/loader/link_neighbor_loader.py:70]: This looks good to me. I think it would be interesting to benchmark the speed here compared to not doing any `unique` call.

Furthermore, you may want to remove the `assert` statement since this will always be true by design. Using `all()` will also call a really slow Python iterator implementation.

Comment (Padarn) [reply torch_geometric/loader/link_neighbor_loader.py:70]: Okay fair enough - I added it incase something changes in the implementation of the base class, but our tests should also catch this.

File: test/loader/test_neighbor_loader.py (hunk 11 above)
Comment (Padarn) [reply test/loader/test_neighbor_loader.py:57]: Actually.. this test fails now 😓  personally I'm not sure if its a problem if this is not true. There could be isolated nodes in the subgraph?

**Review by rusty1s — APPROVED** [review]
Thank you!
